### PR TITLE
Improves linting for issue 503 by adding lint to check ABI has not changed between functions with same `export_name`

### PR DESCRIPTION
**Pull request by ehiggs**
In this PR we check for ABI changes when the export name stayed the same.

Paired with Predrag at FOSDEM. :)

### CODE REVIEW

**Comment by ehiggs** [comment]
Before I updated the branch to latest master, stable was failing; but now 1.71 is failing. I can't reproduce the failure in stable. 

```
❯ uname
Linux
❯ cargo version
cargo 1.75.0 (1d8b05cdd 2023-11-20)

```

**Comment by obi1kenobi** [comment]
Hmm, is it possible we have a flaky test? I'm currently at the airport,
I'll dig in when I get home.

If you have a bit of time, could you re-run the tests a few times and see
if it consistently fails on the same test in the same way or not?

Thanks!

**Comment by ehiggs** [comment]
>I took a look, and the issue is incorrect "expected output" due to one of the functions we renamed shortly before opening the PR:

Yup!

I re-ran the rustdoc generator and fixed the expected results and we have a green light!

Thanks for your help here.

File: src/lints/exported_function_changed_abi.ron [hunk 1]
```diff
@@ -0,0 +1,42 @@
+SemverQuery(
+    id: "function_changed_export_name",
+    human_readable_name: "TODO",
+    description: "TODO",
+    required_update: Major,  // TODO
+    reference_link: None,  // TODO
```
Comment (obi1kenobi) [inline src/lints/exported_function_changed_abi.ron:6]: Suggestion to fill in these TODOs, based on how the other fn ABI lint had them: https://github.com/obi1kenobi/cargo-semver-checks/blob/main/src/lints/function_changed_abi.ron

```suggestion
    human_readable_name: "exported function changed ABI",
    description: "A function marked `#[no_mangle]` or assigned an explicit `#[export_name]` changed its external ABI.",
    required_update: Major,
    reference_link: Some("https://doc.rust-lang.org/reference/items/external-blocks.html#abi"),
```

**Review by obi1kenobi — COMMENTED** [review]
If you think my suggestion about the TODO updates looks good, this is ready to merge! 🚀

Thanks for building this, it was great to pair with you on it.

**Review by obi1kenobi — APPROVED** [review]

**Pull request merged**